python/dex/interop/jax.py

```diff
@@ -189,6 +192,206 @@ def dex_call_batched(batched_args, batched_dims, func_atom):
+  """Evaluates the function output at arg_values, and the linearized function
+  (linearized about arg_values) at arg_tangents.
+
+  Args:
+    arg_values: A tuple of arguments.
+    arg_tangents: A tuple with the tangents of the arguments. The tuple has the
+      same length as the arg_values. Some of the tangents may also be the
+      special value ad.Zero to specify a zero tangent.
+    func_atom: Function atom to linearize. The result type of this function
+      atom must be a single array type.
+
+  Returns:
+     A pair of the primal output and the tangent.
+  """
+  assert len(func_atom.compile().result_signature) == 1
+  num_args = len(arg_values)
+
+  # Helper functions to build strings of primal and tangent inputs.
+  def arg_string(prefix):
+    return " ".join(f"{prefix}{i}" for i in range(num_args))
+
+  def tuple_string(prefix):
+    return "(" + ", ".join(f"{prefix}{i}" for i in range(num_args)) + ")"
+
+  # Add the current function atom as a variable in the context, so that we can
+  # use it to apply batching.
+  env = api.insert(func_atom.module, api.as_cstr("jax_func"), func_atom)
+
+  # `linearize` only seems to work properly for functions which take a single
+  # input argument, so we uncurry `func_atom` to make it into this form. The
+  # evaluated string for three function arguments should look like:
+  # ```
+  # \ (x0, x1, x2). jax_func x0 x1 x2
+  # ```
+  uncurried = eval(
+      r"\ " + tuple_string("x") + ". jax_func " + arg_string("x"),
+      module=env)
+  old_env, env = env, api.insert(env, api.as_cstr("jax_func_uncurried"),
+                                 uncurried)
+  # Destroy the intermediate context.
+  api.destroyContext(old_env)
+
+  # Since we can't return tuple types to Python, we need separate primitives
+  # for the primal and tangent evaluations.
+  #
+  # Here we write out the tangent evaluation expression in pointful style.
+  # The evaluated string for three function arguments should look like:
+  # ```
+  # \ x0 x1 x2 u0 u1 u2.
+  #   linearized = linearize jax_func_uncurried (x0, x1, x2)
+  #   snd linearized (u0 u1 u2)
+  # ```
+  evaluate_linearized = eval(
+      r"\ " + arg_string("x") + " " + arg_string("u") + "."
+      f"\n  linearized = linearize jax_func_uncurried {tuple_string('x')}" +
+      f"\n  snd linearized {tuple_string('u')}",
+      module=env)
+
+  # Materialize jax.ad.Zero values into actual arrays of zeros.
+  tangents_no_zeros = [
+      jax.lax.zeros_like_array(arg) if type(tan) is jax.ad.Zero else tan
+      for arg, tan in zip(arg_values, arg_tangents)
+  ]
+
+  # TODO: Once we can return tuple values from Dex to Python, we might want to
+  # use a single primitive for both forward and linearized evaluation.
+  return (
+      primitive(func_atom)(*arg_values),
+      primitive(evaluate_linearized)(*arg_values, *tangents_no_zeros),
+  )
+
+jax.interpreters.ad.primitive_jvps[dex_call_p] = dex_call_jvp
+
+
+# === transpose ===
```

Comment:
(For the future: when you have self-contained changes, it's better to split them into separate PRs. This makes reviewing much easier.)